Upgrade/Install: Roll back a plugin auto-update with a fatal error.

costdev opened the pull request:
The biggest concern when plugins update automatically is that they may contain a fatal error that crashes a website.

This performs a loopback request to the homepage in an attempt to detect such fatal errors in an active plugin after updating. This ensures that only the plugin's new code has been loaded into memory, and prevents false positives when, for example, a plugin may include a class definition in the main plugin file.

Should a fatal error be detected, the update is restored to the previously installed version.

The loopback request is also performed for the [Plugin File Editor and Theme File Editor](https://github.com/wordpress/wordpress-develop/blob/6a619851c27b4103d7abba5c0a047e3f68c793a6/src/wp-admin/includes/file.php#L378). However, the implementation here does not distinguish between a failed loopback request and a plugin with no fatal errors, as the aim is to make the best attempt to detect a fatal error.

Additionally, the Site Health error message for the loopback request test states:
> Loopback requests are used to run scheduled events, and are also used by the built-in editors for themes and plugins to verify code stability.

So if loopback requests are not possible, automatic updates should not be possible anyway. 🙂

Trac ticket: https://core.trac.wordpress.org/ticket/58281

Commit:
Introduce `WP_Upgrader::fatal_error_detected()`.

Commit:
Roll back a plugin auto-update with a fatal error.

Commit:
Update auto-update email with rollback note.

afragen commented:
@costdev and I have been testing all combinations  of plugins, fataling plugin, and themes with this, even @peterwilsoncc's list of plugins.

It tests out great!

mukeshpanchal27 commented on src/wp-admin/includes/class-wp-upgrader.php:
```diff
@@ -1215,6 +1272,81 @@ public function delete_temp_backup() {
+	protected function fatal_error_detected() {
+		$scrape_key   = md5( rand() );
+		$transient    = 'scrape_key_' . $scrape_key;
+		$scrape_nonce = (string) rand();
```
@costdev could we use `wp_rand` here?

costdev replied:
We could, though this uses [the same approach](https://github.com/wordpress/wordpress-develop/blob/6a619851c27b4103d7abba5c0a047e3f68c793a6/src/wp-admin/includes/file.php#L529) that the Plugin File Editor and Theme File Editor use to generate the transient and nonce. I'm not sure that using `wp_rand()` in this context will provide any benefits.

Commit:
Move checks into `WP_Automatic_Updater`.

Commit:
Add `$temp_backups` parameter to `WP_Upgrader::restore_temp_backup()`.

Commit:
Add `$temp_backups` parameter to `WP_Upgrader::delete_temp_backup()`.

Commit:
Ensure the filesystem is initialized in `WP_Upgrader::maintenance_mode()`.

Commit:
Use `::$temp_backups` in `::delete_temp_backup()`.

Commit:
Don't output feedback during automatic updates.

Commit:
Add logging during scrapes for testing.

Commit:
Move the logic from `::run()` to `::update()`.

This improves integration with Core by:
- Saving on the creation of another instance of `Plugin_Upgrader`.
- Leveraging the "per-plugin" state of `::update()`.
- Utilizing variables already defined in the `::update()` method.

Commit:
Remove potentially inadequate time limit.

This is set to ten minutes in `::update()`.

Commit:
Restore the scrape key timeout to 60 seconds.

Commit:
Add latest updates.

Commit:
Do not enable maintenance mode while scraping.

Commit:
Update maintenance mode locations.

Commit:
Rename `$r` and add an inline comment.

costdev committed:
Harden the maintenance mode bypass while scraping.

costdev committed:
Change hardening approach for maintenance mode bypass.

`get_transient()`, options and the database aren't available this early in the bootstrap.

costdev committed:
Update `@since` tags.

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1551,4 +1678,73 @@ protected function send_debug_email() {
+		}
+
+		$scrape_key    = md5( $upgrading );
+		$scrape_nonce  = (string) $upgrading;
```
```suggestion
		$scrape_nonce  = (string) $upgrading;
		$transient     = 'scrape_key_' . $scrape_key;
		set_transient( $transient, $scrape_nonce, 30 );

		$cookies       = wp_unslash( $_COOKIE );
```

afragen replied:
Not setting the transient results in a scrape_key error as this is checked in `wp_start_scraping_edited_file_errors()`

afragen replied:
Also need to add `$cookies` back.

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1551,4 +1678,73 @@ protected function send_debug_email() {
+	 *
+	 * @return bool Whether a fatal error was detected.
+	 */
+	protected function has_fatal_error( $nonce ) {
```
```suggestion
	protected function has_fatal_error() {
```

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1551,4 +1678,73 @@ protected function send_debug_email() {
+		$needle_start = "###### wp_scraping_result_start:$scrape_key ######";
+		$needle_end   = "###### wp_scraping_result_end:$scrape_key ######";
+		$url          = add_query_arg( $scrape_params, home_url( '/' ) );
+		$response     = wp_remote_get( $url, compact( 'cookies', 'headers', 'timeout', 'sslverify' ) );
```
```suggestion
		$response     = wp_remote_get( $url, compact( 'headers', 'timeout', 'sslverify' ) );
```

afragen replied:
`cookies` not defined and `compact` throwing an error. Likely PHP8+ related.

afragen commented on src/wp-includes/load.php:
```diff
@@ -420,6 +420,16 @@ function wp_is_maintenance_mode() {
 
+	// Don't enable maintenance mode while scraping for fatal errors.
+	if ( is_int( $upgrading ) && isset( $_REQUEST['wp_scrape_key'], $_REQUEST['wp_scrape_nonce'] ) ) {
+		$key   = wp_unslash( $_REQUEST['wp_scrape_key'] );
+		$nonce = wp_unslash( $_REQUEST['wp_scrape_nonce'] );
```
```suggestion
		$key   = stripslashes( $_REQUEST['wp_scrape_key'] );
		$nonce = stripslashes( $_REQUEST['wp_scrape_nonce'] );
```

afragen replied:
`wp_unslash()` not defined yet.

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1551,4 +1678,73 @@ protected function send_debug_email() {
+			$error_output = substr( $error_output, 0, strpos( $error_output, $needle_end ) );
+			$result       = json_decode( trim( $error_output ), true );
+		}
+
```
```suggestion
               delete_transient( $transient );
```

afragen replied:
Should probably add this back too.

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1551,4 +1678,73 @@ protected function send_debug_email() {
+			return false;
+		}
+
+		require_once $maintenance_file;
```
```suggestion
		require $maintenance_file;
```

afragen replied:
need to require every time to load new nonce.

costdev committed:
Apply suggestions from code review

Co-authored-by: Andy Fragen <andy@thefragens.com>

costdev committed:
PHPCS fixes

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -486,6 +520,86 @@ public function update( $type, $item ) {
+				// Avoid a race condition when there are 2 sequential plugins that have fatal errors.
+				sleep( 2 );
+
+				if ( $this->has_fatal_error() ) {
+					$upgrade_result = new WP_Error();
```
```suggestion
				$has_fatal_error = $this->has_fatal_error();
				if ( $has_fatal_error ) {
					$upgrade_result = is_wp_error( $has_fatal_error ) ? $has_fatal_error : new WP_Error();
```

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1551,4 +1678,77 @@ protected function send_debug_email() {
+	 *
+	 * @global int $upgrading The Unix timestamp marking when upgrading WordPress began.
+	 *
+	 * @return bool Whether a fatal error was detected.
```
```suggestion
	 * @return bool|WP_Error Whether a fatal error was detected. WP_Error if loopback not functioning.
```

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1551,4 +1678,77 @@ protected function send_debug_email() {
+		$url          = add_query_arg( $scrape_params, home_url( '/' ) );
+		$response     = wp_remote_get( $url, compact( 'cookies', 'headers', 'timeout', 'sslverify' ) );
+
+		// If this outputs `true` in the log, it means there were no fatal errors detected.
```
```suggestion
		if ( is_wp_error( $response ) ) {
			$response->add_data( 'Loopback is having a problem and any auto-updates will be rollback for safety.' );

			return $response;
		}

		// If this outputs `true` in the log, it means there were no fatal errors detected.
```

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1551,4 +1678,77 @@ protected function send_debug_email() {
+		$needle_end   = "###### wp_scraping_result_end:$scrape_key ######";
+		$url          = add_query_arg( $scrape_params, home_url( '/' ) );
+		$response     = wp_remote_get( $url, compact( 'cookies', 'headers', 'timeout', 'sslverify' ) );
+
```
```suggestion

		if ( is_wp_error( $response ) ) {
			error_log( 'Loopback request failed: ' . $response->get_error_message() );
			return true;
		}

```

afragen replied:
Hopefully this should add the loopback error message and cause the auto-update to rollback. It integrates better into the existing code.

afragen replied:
This tests out fine on my end. I intentionally caused a `WP_Error` in the HTTP Request.

```
[10-Apr-2024 18:38:18 UTC] Automatic updates starting...
[10-Apr-2024 18:38:19 UTC]   Automatic plugin updates starting...
[10-Apr-2024 18:38:19 UTC]     Upgrading plugin 'this-plugin-should-not-be-used'...
[10-Apr-2024 18:38:35 UTC]     Plugin 'this-plugin-should-not-be-used' has been upgraded.
[10-Apr-2024 18:38:37 UTC]     Scraping home page...
[10-Apr-2024 18:39:03 UTC] '###### wp_scraping_result_start:aef75992abd5bdd8c7491a1f558742c0 ######
{"type":"1","message":"Loopback is having a problem and any auto-updates will be rollback for safety."}
###### wp_scraping_result_end:aef75992abd5bdd8c7491a1f558742c0 ######'
[10-Apr-2024 18:39:03 UTC]     The update for 'this-plugin-should-not-be-used' contained a fatal error. The previously installed version has been restored.
[10-Apr-2024 18:39:03 UTC]   Automatic plugin updates complete.
[10-Apr-2024 18:39:03 UTC]   Automatic theme updates starting...
[10-Apr-2024 18:39:03 UTC]   Automatic theme updates complete.
[10-Apr-2024 18:39:03 UTC] Automatic updates complete.
```

costdev replied:
It seems like this could be a shorthand of the above:

```php
if ( is_wp_error( $response ) ) {
    error_log( 'Loopback is having a problem and the update will be rolled back for safety.' );
    return true;
}
```

We can uncouple the message from rollback, and ensure the error from the response is included in the output, such as:

```php
error_log( 'Loopback request failed: ' . $response->get_error_message() );
```

afragen replied:
I understand the first part and yes it would be a shortcut. 

I don’t understand how the uncoupling works or is that a replacement for the shortcut’s error message?

costdev replied:
Yeah that's right, it's a replacement for the error message.

afragen replied:
Let's go with the shortcut and finish off the merge proposal so we can watch all the testing being done in the following 2 weeks and then commit if no issues reported.

afragen replied:
Don’t we need to report that the update was rolled back?

costdev replied:
Yep, but only in the _usage_ of `has_fatal_error()` (when the rollback is performed) and in the update email.

afragen replied:
The shortcut won't have any messaging about the rollback. Only the email.

costdev replied:
That's right yeah. The method's used as a component of Rollback Auto-Update, but it's a useful piece of functionality on it's own, so we should try to remove mention of Rollback inside it so it has potential for use elsewhere.

Rather than mention rolling back in `::has_fatal_error()`, which is about trying to detect a fatal error, we instead already mention rolling back in `::update()` via the `WP_Error` messages currently in the PR, where we act based on the result of `::has_fatal_error()`.

afragen replied:
Suggestion updated to correspond to ongoing discussion.

afragen reviewed (changes requested):
This should better integrate without the need to return a WP_Error

costdev committed:
Simplify the return for loopback failure.

Co-authored-by: Andy Fragen <andy@thefragens.com>

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1551,4 +1678,81 @@ protected function send_debug_email() {
+			$error_output = substr( $error_output, 0, strpos( $error_output, $needle_end ) );
+			$result       = json_decode( trim( $error_output ), true );
+		}
+				delete_transient( $transient );
```
I think there are extra tabs here

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1551,4 +1678,81 @@ protected function send_debug_email() {
+		if ( is_wp_error( $response ) ) {
+			error_log( 'Loopback request failed: ' . $response->get_error_message() );
+			return true;
+		}
```
Maybe a return after this line?

costdev committed:
Improve spacing.

audrasjb commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -486,6 +520,86 @@ public function update( $type, $item ) {
+							sprintf(
+								/* translators: %s: The plugin's slug. */
+								__( "The update for '%s' contained a fatal error. The previously installed version could not be restored." ),
+								$item->slug
```
Maybe it would be better to also display the full plugin name?
Something like `The update for Plugin Name (plugin-slug) contained a fatal error. The previously installed version could not be restored.`

costdev replied:
Makes sense! I'll update the PR a little later

audrasjb replied:
👍

afragen replied:
Only concern I see with this is the length of some of the plugin names. As this is only in the error log I think the slug makes more sense.

costdev replied:
This error actually gets added to the `$update_results['plugin']` array rather than being in the error log.

Even still, some plugins do have very long names. For example, not to single it out, it just comes to mind:

> WPSSO Core – The Complete Structured Data Plugin: Schema Markup, Social and SEO Meta Tags, Snippets, Feeds, Sitemaps, Breadcrumbs

@audrasjb What do you think?

audrasjb replied:
Yeah that's right. Let's start with the slug. As we're dealing with logs, let's start with this technicval information only 👍

audrasjb commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -486,6 +520,86 @@ public function update( $type, $item ) {
+							'plugin_update_fatal_error_rollback_successful',
+							sprintf(
+								/* translators: %s: The plugin's slug. */
+								__( "The update for '%s' contained a fatal error. The previously installed version has been restored." ),
```
Maybe it would be better to also display the full plugin name?
Something like `The update for Plugin Name (plugin-slug) contained a fatal error. The previously installed version has been restored.`

audrasjb commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -486,6 +520,86 @@ public function update( $type, $item ) {
+					 */
+					error_log( '    ' . implode( "\n", $upgrade_result->get_error_messages() ) );
+				} else {
+					error_log( '    The update for ' . var_export( $item->slug, true ) . ' has no fatal errors.' );
```
I'm not sure we should log anything when everything went fine :)

costdev replied:
This is just for testing purposes so that testers can see and verify the process more easily.

audrasjb replied:
yeah, thanks :)

audrasjb commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -524,24 +640,35 @@ public function run() {
 		wp_update_plugins(); // Check for plugin updates.
 		$plugin_updates = get_site_transient( 'update_plugins' );
 		if ( $plugin_updates && ! empty( $plugin_updates->response ) ) {
+			error_log( '  Automatic plugin updates starting...' );
```
Same here: do we need to log everything that happens during the update process?

costdev replied:
Same as above, just for helping testers. :slightly_smiling_face:

audrasjb replied:
Yeah I just saw this in the Make/Core blogpost 👍

audrasjb reviewed (commented):
I added a few comments :)

justlevine commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -1551,4 +1678,84 @@ protected function send_debug_email() {
+
+		$needle_start = "###### wp_scraping_result_start:$scrape_key ######";
+		$needle_end   = "###### wp_scraping_result_end:$scrape_key ######";
+		$url          = add_query_arg( $scrape_params, home_url( '/' ) );
```
Can we make the `$url`  filterable? Many plugins don't affect the frontpage (or shared lifecycle) and won't be caught by a naive check. Adding a filter here would give plugin developers a way to ensure/provide a relevant rollback check. E.g.

- A utility plugin checking it's Admin screen loads.
- A CPT plugin checking the CPT archive doesn't fatal.
- A large ecosystem plugin checking a custom endpoint that runs a series of post-update health checks.

costdev replied:
> Can we make the $url filterable?

That's actually something we considered during the feature's earlier development. It was removed as we wanted to start with the core feature and iterate as requests were made. I'm happy to add the filter, but I'll leave this conversation open for a bit first to see if there is more support for it, or possibly some opposition to it at this time.

>- A utility plugin checking it's Admin screen loads.
> - A large ecosystem plugin checking a custom endpoint that runs a series of post-update health checks.

Loopback requests for admin will require authentication, so this is likely not possible with the current setup (and I _believe_ that's the case with the Plugin File Editor and Theme File Editor too). It's certainly something we could look at in a future iteration, though I think it might require somewhat of a deep-dive to look at implementation options.

> - A CPT plugin checking the CPT archive doesn't fatal.

Yep, that would be achievable with a filter, assuming the CPT archive is a public page (usually is, but some may be "members-only" areas).

justlevine replied:
Thanks @costdev !

> I'm happy to add the filter, but I'll leave this conversation open for a bit first to see if there is more support for it, or possibly some opposition to it at this time.

I'll share in `#outreach` to get some more voices to weigh in 🤞 

>> A utility plugin checking it's Admin screen loads.
> A large ecosystem plugin checking a custom endpoint that runs a series of post-update health checks.

>Loopback requests for admin will require authentication, so this is likely not possible with the current setup (and I believe that's the case with the Plugin File Editor and Theme File Editor too).

Great catch! While it wouldnt be too complicated to make the rest of the `wp_remote_get()` params filterable, that definitely is overkill for an initial implementation. Besides, utility plugins could also take the approach of using a custom endpoint, even if that's a bit more work than just checking an admin screen they've already developed/hooked into.

afragen replied:
The goal is to prevent a site from a PHP fatal after a plugin update. That typically means checking the homepage. If some custom endpoint results in a PHP fatal will most users even see that?

costdev replied:
@afragen touches on something here: Our main targets during development were fatal errors that would completely render a site unusable, rather than say, checking a plugin's functionality. Not that a single page type fataling isn't bad - of course it is - but our main focus at this stage has been ones where the homepage can't load, much like the Plugin File Editor and Theme File Editor, and the file edited in those editors could be for any page/template/endpoint.

Also note that a filtered URL would _replace_ the URL to be scraped, rather than add another URL to scrape. That means a plugin's URL may work, while the homepage doesn't. We can't scrape multiple URLs because that would cause an unknown delay to every update.

Where I'm at currently (pending further discussion): For consistency with the Plugin File Editor and Theme File Editor, as well as ensuring the homepage is protected, I'm leaning towards keeping the homepage as the only URL that's scraped, at least for this initial version of the feature. Enhancing too early can risk undermining the feature, and since this is a stability feature, that'd be counter-productive.

justlevine replied:
@afragen 

>The goal is to prevent a site from a PHP fatal after a plugin update. That typically means checking the homepage.If some custom endpoint results in a PHP fatal will most users even see that?

"Typically" or currently in this PR and an separate feature that's similar from a code perspective but not from a usage perspective?
If Co-Authors Plus breaks on a product site, it's the blog page that will _always_ fatal, while the home page only will fatal if theres a recent posts section.
Same thing with a multipurpose site using WooCommerce+ the Shop endpoint. An extension like Woo Payments or  WooCommerce [x] Gateway would naively affect the checkout page, and more likely fatal during the checkout _process_, which is where a custom health check endpoint could be used if the plugin author felt it was worthwhile.

(@costdev , I believe most would consider the above examples as "fatal errors that would completely render a site unusable", from either an end user or developer perspective)

@costdev 
> Where I'm at currently (pending further discussion): For consistency with the Plugin File Editor and Theme File Editor, as well as ensuring the homepage is protected, I'm leaning towards keeping the homepage as the only URL that's scraped, at least for this initial version of the feature.

Using the Plugin/Theme File Editor as the measure of consistency seems like the wrong benchmark IMO.
- From an engineering POV: at a glance, it doesn't look like this feature is sharing code with the Plugin/Theme checker, so there's no technological need for consistency.
- From an end-user POV, the use cases are significantly different:
  - If you're actively making changes to an existing plugin, then you (ostensibly) know how to test what what you're changing. In this scenario, the Fatal Error checker is a quick, extra safeguard to make sure you didn't  accidentally do something obviously broken (that would otherwise have been caught by your IDE/CI if you weren't making code changes via the browser). More targeted health checks are wholly additive, don't affect the goal of the feature, and offer limited benefit.
  - The goal of Plugin Rollbacks (as I understand it) is provide some form of protection against incompatible code changes by a third party, and even more so in an auto-update scenario where youre not even aware of these changes until after the fact. As such a naive homepage check doesn't provide an additive guard, but rather a limited and often _ineffective_ guard.
  In the best case scenario, the feature limitations are clearly communicated and users stay opted-out of Auto-Updates for most of their plugins; worst case, the false expectations have users turning on Auto-Updates for more plugins, they still get updates that break their site, and are even more trepidatious about Auto Updates in the future.
  
> Enhancing too early can risk undermining the feature, and since this is a stability feature, that'd be counter-productive.



Per the above, I believe the risk here is reversed. From a user perspective, an inneffectice guard is more "unstable" (in context: "unreliable") than.... What exactly?

As a reminder, we're talking about plugins additively being able to provide a different string URI (that's more relevant than the homepage) to loopback instead of the default hard-coded `/`. The minimal API footprint is the URI string and the plugin identifier. What are some cases you worried about that would undermine the feature?

costdev replied:
_Before reading_: I'm on mobile and between tasks, so hopefully my brain -> keyboard coordination holds up here :sweat_smile:

> I believe most would consider the above examples as "fatal errors that would completely render a site unusable", from either an end user or developer perspective

Though I'm not talking about a subjective interpretation of a site being unusable, I might even agree with every example you've given. I'm talking about when a site's homepage doesn't load, which may indicate a more fundamental browsability issue. i.e. If a homepage fatals, the site is unusable by everyone trying to view that site's top-level URL, the main arrival point and "base", and there's nowhere to go from there. If a sub-page breaks, the site is still browsable from its top-level URL. A sub-page may work while the homepage doesn't, but who can reach the sub-page to find out?

For example, WooCommerce could choose one of many endpoint URLs where something could go wrong and it's critical to the plugin's functionality. If it picked one, I imagine it would be the shop page as its main entry point (i.e. no point verifying a later point you can't reach because the shop page isn't even working). It would be limited and possibly ineffective, but it'd at least be trying to protect users at its entry point. That's akin to what we're doing here, as WordPress' entrypoint. Even if other URL checks were desired, a check on the shop page shouldn't be replaced, and that's again where I'm hesitant to introduce a filter that would replace our check on the homepage; a site's main entry point.

You mentioned about a plugin that might only affect a blog page, though for example, any plugin intended to have its "own space" may hook `init`, and contain a conflict and fatal error in the callback that runs on every page, thus objectively rendering _the whole site_ unusable. That's the kind of fundamental issue I mean.

---

> From an engineering POV: at a glance, it doesn't look like this feature is sharing code with the Plugin/Theme checker, so there's no technological need for consistency.

This isn't my response but just an FYI: The feature uses the same (not a copy) scraping code as the Plugin File Editor and Theme File Editor. While the request is its own in `::has_fatal_error()`, the actual scraping is the same code.

While we're aiming for consistency at an engineering level too, the consistency I referred to earlier isn't necessarily at an engineering level, but that the homepage is checked, and also the thought process.

From [the commit](https://core.trac.wordpress.org/changeset/41721) that introduced these editors:
> Introduce sandboxed PHP file edits for active themes, preventing accidental whitescreening of a user's site when introducing a fatal error.
> After writing a change to a PHP file for an active theme or plugin, perform loopback requests on the file editor admin screens and the homepage to check for fatal errors. If a fatal error is encountered, roll back the edited file and display the error to the user to fix and try again.

And from the discussion on one of the [related tickets](https://core.trac.wordpress.org/ticket/21622#comment:18):

> we can at least make sure the front page still loads to mitigate some forms of breakage is my thinking.

---

> If you're actively making changes to an existing plugin, then you (ostensibly) know how to test what what you're changing. In this scenario, the Fatal Error checker is a quick, extra safeguard to make sure you didn't accidentally do something obviously broken (that would otherwise have been caught by your IDE/CI if you weren't making code changes via the browser). More targeted health checks are wholly additive, don't affect the goal of the feature, and offer limited benefit.

The fatal error checker for the editors is quite critical. It prevents you editing a file such that you may not be able to access the same admin screen to repair the issue, and performs an additional check to the homepage such that the site is still browsable. These editors are often used by users of all ranges, from experienced developers with no FTP access, to site owners with no code experience who are using snippets they found on a blog post, so whether they know exactly how to test (fully) the potential impact of their changes is 50/50.

---

> The goal of Plugin Rollbacks (as I understand it) is provide some form of protection against incompatible code changes by a third party, and even more so in an auto-update scenario where youre not even aware of these changes until after the fact. As such a naive homepage check doesn't provide an additive guard, but rather a limited and often ineffective guard.
In the best case scenario, the feature limitations are clearly communicated and users stay opted-out of Auto-Updates for most of their plugins; worst case, the false expectations have users turning on Auto-Updates for more plugins, they still get updates that break their site, and are even more trepidatious about Auto Updates in the future.


As demonstrated earlier in this thread, I do see the value in a more targeted check by plugins, but not as a replacement for a homepage check. I can hope a filter would only be used when a plugin author is positive that their plugin's code doesn't run on the homepage, or during any bootstrapping/init, but historically, intended usage and actual usage differ and have a bunch of use cases in between. What we aim to guarantee with Rollback Auto-Update, at least in this initial version, is that an automatic update will not present visitors to your homepage with a WSOD that may indicate issues that could be present on every other page of your site.

As an aside and a personal preference, if it's okay with you I'd prefer not talk in terms of best/worst case scenarios while discussing a disagreement, as it can often result in a back and forth that eventually (even surprisingly) becomes emotive, with people nit-picking at each other's speculation that's required for best/worst cases, and would rather continue our discussion as we have earlier in the thread. It just tends to be a better discussion, at least in my opinion/experience.

---

> Per the above, I believe the risk here is reversed. From a user perspective, an inneffectice guard is more "unstable" (in context: "unreliable") than.... What exactly?

Unfortunately (and I hate it), we can never be 100% reliable against fatal errors as we can't execute all code paths in a plugin. What we can do is make sure that the site still loads more than a WSOD at its entrypoint. IMO, that should be a _base_ check, with any others being additional scrapes - and that's where the problem with a filter lies for me, either it replaces the homepage check, or it's additional and could put the whole update process at risk of timeout, etc.

---

> As a reminder, we're talking about plugins additively being able to provide a different string URI (that's more relevant than the homepage) to loopback instead of the default hard-coded /. The minimal API footprint is the URI string and the plugin identifier. What are some cases you worried about that would undermine the feature?

I mentioned above about an `init` hook's callback containing a conflict that occurs on every page, and a plugin author can't reasonably be expected to have tested against all plugins for a conflict.

Another case would be that a hook that should run on every page assumes the availability of a plugin's functionality, and it only worked when a plugin author tested it because of how they'd configured their site. In other words, a case where "happy path"-only testing strikes again.

---

With all the above said, I like the idea of a plugin being able to check that a particular page can load, but not as a replacement of the homepage check. I just can't see a way to accomplish that without impacting the update process' execution. That may just be a limitation of my imagination at the moment, I don't know, but that's why I want to keep discussing this :slightly_smiling_face:

justlevine replied:
Thanks @costdev so much for the attention and detail - especially from mobile!

Before anything else:
> As an aside and a personal preference, if it's okay with you I'd prefer not talk in terms of best/worst case scenarios while discussing a disagreement, as it can often result in a back and forth that eventually (even surprisingly) becomes emotive.

This is a lovely and appreciated tip. Thank you 💙 

> I'm talking about when a site's homepage doesn't load, which may indicate a more fundamental browsability issue. i.e. If a homepage fatals, the site is unusable by everyone trying to view that site's top-level URL, the main arrival point and "base", and there's nowhere to go from there. If a sub-page breaks, the site is still browsable from its top-level URL. A sub-page may work while the homepage doesn't, but who can reach the sub-page to find out?

_This_ is the important context I was missing that brings everything else into focus. I was looking at this feature as a guard against plugins breaking _on_ the site, and not plugins breaking _access to_ the site. From the latter perspective, we indeed wouldn't want to replace the `home_url` check, and while it doesn't address the larger concerns about auto-update stability that sort of work makes much more sense as possible enhancements only after this more fundamental check is in place 👍

Responding to a couple of your individual points for future reference (either others joining this discussion, or prior art review when its time to start talking future possible enhancements).

---

> The fatal error checker for the editors is quite critical. It prevents you editing a file such that you may not be able to access the same admin screen to repair the issue, and performs an additional check to the homepage such that the site is still browsable.

Understood. It was for this reason I suggested the use case and thus the guard required are different. Users aren't editing code when updating a plugin, or taking _any_ manual actions during Auto-Updates. Ensuring users don't lock themselves out of being able to revert the change doesn't apply (unless/until a rollback GUI is introduced), and from a _broken plugin_ POV the homepage check isn't a strong indicator of anything. From a _site access_ POV, that homepage check still resonates.

> I do see the value in a more targeted check by plugins, but not as a replacement for a homepage check. I can hope a filter would only be used when a plugin author is positive that their plugin's code doesn't run on the homepage, or during any bootstrapping/init, but historically, intended usage and actual usage differ and have a bunch of use cases in between.

Agreed - seems like future enhancements to the feature would be better served by an action (or filter that defaults to `null`) that allows plugins to run their own additional health check, without replacing the home_url access check.

johnbillion commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -486,6 +520,86 @@ public function update( $type, $item ) {
+				 */
+				if ( function_exists( 'set_time_limit' ) ) {
+					set_time_limit( 10 * MINUTE_IN_SECONDS );
+				}
```
Can this call to `set_time_limit()` be moved into the `has_fatal_error()` method so it's nearer the loopback request? This is what happens in the existing `wp_edit_theme_plugin_file()` function.

costdev replied:
It can, although I added it here so that it would be in the context of both the loopback request and the potential rollback. It's also possible that the `::has_fatal_error()` method may be used by extenders who won't want such a long timeout to be set, though I'm not sure that's much of a concern. I don't have any strong opinions on where this is located though, so if you still think it's best placed in `::has_fatal_error()`, I'm happy to relocate it.

johnbillion replied:
I don't have a strong opinion either. Let's leave it here.

johnbillion commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -486,6 +520,86 @@ public function update( $type, $item ) {
+				}
+
+				// Avoid a race condition when there are 2 sequential plugins that have fatal errors.
+				sleep( 2 );
```
Is there further information that can be added to the comment above this magic number?

costdev replied:
@afragen Since this was added as a result of your testing, you may be best placed to elaborate on this.

afragen replied:
In previous testing it was possible that rollbacks overlapped. This pause is also used in `move_dir` due to an issue in VirtualBox. 

It came about when testing against multiple plugin fatals requiring rollback. 

https://github.com/WordPress/rollback-update-failure/commit/9ed09ea855cec8e3d733702c168bca6625412c1b

johnbillion commented on src/wp-admin/includes/class-wp-upgrader.php:
```diff
@@ -1129,17 +1144,29 @@ public function move_to_temp_backup_dir( $args ) {
 	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
 	 *
+	 * @param array[] $temp_backups {
+	 *      Optional. An array of temporary backups.
+	 *
+	 *      @type string $dir  The temporary backup location in the upgrade-temp-backup directory.
+	 *      @type string $slug The item's slug.
+	 *      @type string $src  The directory where the original is stored. For example, `WP_PLUGIN_DIR`.
```
```suggestion
	 *     Optional. An array of temporary backups.
	 *
	 *     @type array ...$0 {
	 *         Information about the backup.
	 *
	 *         @type string $dir  The temporary backup location in the upgrade-temp-backup directory.
	 *         @type string $slug The item's slug.
	 *         @type string $src  The directory where the original is stored. For example, `WP_PLUGIN_DIR`.
	 *     }
```

johnbillion commented on src/wp-admin/includes/class-wp-upgrader.php:
```diff
@@ -1182,17 +1209,29 @@ public function restore_temp_backup() {
 	 * @global WP_Filesystem_Base $wp_filesystem WordPress filesystem subclass.
 	 *
+	 * @param array[] $temp_backups {
+	 *      Optional. An array of temporary backups.
+	 *
+	 *      @type string $dir  The temporary backup location in the upgrade-temp-backup directory.
+	 *      @type string $slug The item's slug.
+	 *      @type string $src  The directory where the original is stored. For example, `WP_PLUGIN_DIR`.
```
```suggestion
	 *     Optional. An array of temporary backups.
	 *
	 *     @type array ...$0 {
	 *         Information about the backup.
	 *
	 *         @type string $dir  The temporary backup location in the upgrade-temp-backup directory.
	 *         @type string $slug The item's slug.
	 *         @type string $src  The directory where the original is stored. For example, `WP_PLUGIN_DIR`.
	 *     }
```

johnbillion commented:
I added some minor code review comments, but in general it's looking good. I'm going to functionally test this over the next few days.

costdev committed:
Update src/wp-admin/includes/class-wp-upgrader.php

Co-authored-by: John Blackbourn <johnbillion@gmail.com>

costdev committed:
Update src/wp-admin/includes/class-wp-upgrader.php

Co-authored-by: John Blackbourn <johnbillion@gmail.com>

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -486,6 +520,86 @@ public function update( $type, $item ) {
+					set_time_limit( 10 * MINUTE_IN_SECONDS );
+				}
+
+				// Avoid a race condition when there are 2 sequential plugins that have fatal errors.
```
```suggestion
				/*
				 * Avoids a race condition when there are 2 sequential plugins that have fatal errors.
				 * WP_Automatic_Updater::update() can be called for the second plugin before the first has
				 * completed WP_Automatic_Updater::has_fatal_error() skipping has_fatal_error() for the
				 * second, and possibly subsequent plugins.
				 */
```

afragen replied:
@johnbillion I think this has a better explanation.

costdev commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -486,6 +520,86 @@ public function update( $type, $item ) {
+					set_time_limit( 10 * MINUTE_IN_SECONDS );
+				}
+
+				// Avoid a race condition when there are 2 sequential plugins that have fatal errors.
```
Slight modification:
```suggestion
				/*
				 * Avoids a race condition when there are 2 sequential plugins that have fatal errors.
				 * WP_Automatic_Updater::update() can be called for the second plugin before the first has
				 * finished checking for fatal errors. This can cause the second plugin's fatal error checking to be skipped,
				 * and may affect subsequent plugins too.
				 */
```

afragen replied:
@johnbillion what do you think of this one?

afragen replied:
I think I still need to tweak this a bit. After testing locally with sequentially fataling plugins I found that without a 2 sec delay before the loopback check, the check does not fail, I assume it has something to do with loading the new plugin into memory rather than anything else. Very strange, but at least the length of time for the loopback shouldn't be an issue.

I tried using `wp_opcache_invalidate_directory()` instead of the sleep but that failed too.

I also tried a `sleep(1)` but that failed.

Looks like `sleep(2)` it is.

johnbillion committed:
Update src/wp-admin/includes/class-wp-automatic-updater.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

afragen commented on src/wp-admin/includes/class-wp-automatic-updater.php:
```diff
@@ -486,6 +520,91 @@ public function update( $type, $item ) {
+					set_time_limit( 10 * MINUTE_IN_SECONDS );
+				}
+
+				/*
+				 * Avoids a race condition when there are 2 sequential plugins that have fatal errors.
+				 * WP_Automatic_Updater::update() can be called for the second plugin before the first has
+				 * finished checking for fatal errors. This can cause the second plugin's fatal error checking to be skipped,
+				 * and may affect subsequent plugins too.
+				 */
```
```suggestion
				/*
				 * Avoids a race condition when there are 2 sequential plugins that have fatal errors.
				 * It seems a slight delay is required for the loopback to use the updated plugin code in the request.
				 * This can cause the second plugin's fatal error checking to be inaccurate,
				 * and may also affect subsequent plugin checks.
				 */
```

costdev replied:
I'm not sure that documenting what doesn't work is useful here, but posting on the ticket what's been checked and verified is worth doing for posterity. In the code, we should document the reasoning for the sleep, and ensure that we've verified the details as much as we can.

afragen replied:
By all means modify as necessary.

afragen replied:
Updated

johnbillion commented:
What shall we do with all the `error_log()` calls in this PR? Strip them out?

johnbillion committed:
Merge branch 'trunk' into rollback_automatic_plugin_updates

costdev commented:
For debugging purposes, since this is a background task, it would be great if we could put them behind a `WP_DEBUG` flag so any reported issues can detail where things went wrong, then remove this logging later in the cycle.

afragen commented:
I say leave them for now for broader testing. Somewhere around beta1 we can put the loopback results behind WP_DEBUG and remove most of the others.

johnbillion commented:
https://core.trac.wordpress.org/changeset/58128 🚀
